src/tplr/comms.py

```diff
@@ -1371,27 +1487,66 @@ async def load_checkpoint(
         bool, dict, int, torch.optim.Optimizer, torch.optim.lr_scheduler._LRScheduler
     ]:
         """
-        Loads the latest checkpoint. No catchup or step simulation happens here.
+        Loads the latest checkpoint. Rank 0 loads, broadcasts, others receive.
+        This function itself *only* loads on Rank 0. Broadcasting happens externally.
         Returns:
-            tuple: (success: bool, momentum: dict, checkpoint_current_window: int,
+            tuple: (load_success_on_rank0: bool, state_data_for_broadcast: dict, checkpoint_sync_window: int,
                     optimizer: Optimizer, scheduler: LRScheduler)
         """
+        if not distrib.is_rank0():
+            return False, {}, 0, optimizer, scheduler
+
+        tplr.logger.info("[Rank 0] Attempting to load checkpoint...")
+        device = f"cuda:{self.local_rank}" if torch.cuda.is_available() else "cpu"
+
         init_version = init_version if init_version is not None else __version__
         result = await self.get_latest_checkpoint(init_version)
         if not result:
-            tplr.logger.info("No valid checkpoints found")
+            tplr.logger.info(
+                "[Rank 0] No valid checkpoints found via get_latest_checkpoint."
+            )
             return False, {}, 0, optimizer, scheduler
 
         checkpoint_data, checkpoint_window = result
         try:
             # 1) Load model and optimizer state
-            model.load_state_dict(
-                {
-                    k: v.to(device)
-                    for k, v in checkpoint_data["model_state_dict"].items()
-                }
+            ckpt_sd: dict[str, torch.Tensor] = checkpoint_data["model_state_dict"]
+
+            # handle mismatched prefixes -------------------------------------------
+            is_wrapped = isinstance(model, torch.nn.parallel.DistributedDataParallel)
+
+            # Get first key from checkpoint to determine its prefix structure
+            first_key = next(iter(ckpt_sd))
+
+            # Remove any DDP prefix from checkpoint keys to match the actual model structure
+            if first_key.startswith("module.model."):
```

Comment:
this seems like a fragile hack. can we have another look how to solve it in the source?